runtime/jcl/common/jclcinit.c

```diff
@@ -712,6 +717,35 @@ initializeRequiredClasses(J9VMThread *vmThread, char* dllName)
 		}
 	}
 
+	if (J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_LOAD_AGENT_MODULE)
+			&& (NULL != vm->modulesPathEntry->extraInfo)) {
+		Trc_JCL_initializeRequiredClasses_addAgentModuleEntry(vmThread);
+		const char *module = vm->jimageIntf->jimagePackageToModule(
+				vm->jimageIntf, (UDATA) vm->modulesPathEntry->extraInfo,
+				"jdk/internal/agent");
+		if (NULL != module) {
+			J9VMSystemProperty *systemProperty = NULL;
+			/* Handle the case where there are no user-specified modules. In this case, there
+			 * is typically one system property but no user-set "add-modules" arguments.
+			 */
+			if ((0 == vm->addModulesCount) &&
+					(J9SYSPROP_ERROR_NONE == vm->internalVMFunctions->getSystemProperty(vm, ADDMODS_PROPERTY_BASE "0", &systemProperty))
+					) {
+				/* this is implicitly an unused property */
+				vm->internalVMFunctions->setSystemProperty(vm, systemProperty, AGENT_MODULE_NAME);
+			} else {
+				UDATA indexLen = (UDATA)floor(log10((double)vm->addModulesCount) + 1);
+				char *propNameBuffer = j9mem_allocate_memory(sizeof(ADDMODS_PROPERTY_BASE) + indexLen, OMRMEM_CATEGORY_VM);
+				if (NULL == propNameBuffer) {
+					Trc_JCL_initializeRequiredClasses_addAgentModuleOutOfMemory(vmThread);
+					return 1;
```

Comment:
Not a fan of returns in the middle of the code. I greatly prefer setting an rc and using goto.

Reply:
> Not a fan of returns in the middle of the code. I greatly prefer setting an rc and using goto.

Me too, but I am following the pattern used throughout the function for consistency.  Let me know if you want me to change to the other instances.

Reply:
Fair enough - let's leave it for now.